File: 0150_evaluate_reverse_polish_notation/solution02.py
import math 
from typing import List 

# 29 Jan 2025
# runtime: 9.07%, memory: 100.00% 
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        oper_funcs = {
            "+": lambda x, y : x + y,
            "-": lambda x, y : x - y,
            "*": lambda x, y : x * y,
            "/": lambda x, y : math.trunc(x / y),
        }
        stack = []

        for t in tokens:
            if t in oper_funcs.keys():
                op2 = int(stack.pop())
                op1 = int(stack.pop())
                t = f"{oper_funcs[t](op1, op2)}"

            stack.append(t)

        return int(stack.pop())

        

File: 0150_evaluate_reverse_polish_notation/solution02.cpp
#include <cstdlib>
#include <stack>

// 20 Mar 2026 
// runtime: 100.00%, memory: 66.07% 
class Solution {
public:
    int operate(int x, int y, const string& op) {
        if (op == "+") 
            return x + y;
        else if (op == "-")
            return x - y;
        else if (op == "*") 
            return x * y;
        else // op == "/"
            return x / y;
    }

    int evalRPN(vector<string>& tokens) {
        stack<int> st;

        for (const string tok : tokens) {
            if (tok == "+" || tok == "-" || tok == "*" || tok == "/") {
                int y = st.top(); st.pop();
                int x = st.top(); st.pop();

                st.push(operate(x, y, tok));
            } else {
                st.push(stoi(tok));
            }
        }

        return st.top();
    }
};